File: rubyserver/lib/algorithms.rb
module Algorithms

   def self.swap(arr, x, y)
   	  _temp = arr[x]
   	  arr[x] = arr[y]
   	  arr[y] = _temp
   	  return arr
   end	

   def self.insertion(arr) 
   	
      arr.each_index do |i|
      	j = i
         #for (j = i; j > 0; j--)
      	while j > 0 
            if (arr[j] < arr[j-1])
               swap(arr, j, j-1)
            else
               break
            end
            j -= 1
         end
      end

      return arr
   end 

   def self.selection(arr)
      arr_length = arr.length

      #for (var i = 0; i < N; i++) {
      arr.each_index do |i|
         min = i;
         #for (var j = i+1; j < N; j++) {
         j = i+1
         while j < arr_length
            if (arr[j] < arr[min])
               min = j;
            end
            j += 1   
         end
         swap(arr, i , min)      
      end

      return arr
   end 

end

File: src/go/tekst/algorithms.go
package tekst

import (
    "log"
    "sort"
)

func swap (arr []string, x, y int) {
    var temp = arr[x]
    arr[x] = arr[y]
    arr[y] = temp
}

func NativeSort(arr []string) (output []string, cnt int) {
    cnt = len(arr)
    sort.Strings(arr)
    output = arr
    log.Println(output)
    return
}

func SelectionSort(arr []string) (output []string, cnt int) {
    cnt = len(arr)
    var min int

    for i := 0; i < cnt; i++ {
        min = i;

        for j := i+1; j < cnt; j++ {
            if (arr[j] < arr[min]) {
                min = j;
            }
        }
        //swap the min found with pointer i
        swap(arr, i , min);
    }
    output = arr
    return
}

func InsertionSort(arr []string) (output []string, cnt int) {
    cnt = len(arr)

    for i := 0; i < cnt; i++ {
        for j := i; j > 0; j-- {
            if (arr[j] < arr[j-1]) {
                swap(arr, j, j-1);
            } else {
                break
            }
        }
    }
    output = arr
    return
}
